In [41]:
import pandas
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error


df = pandas.read_csv('Datasets/winequality-red.csv',sep=";")
df


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1599 rows x 12 columns]

# Entrenamiento y prueba

In [42]:


X = df.values[:,:-1]
Y = df.values[:,-1]
Xtrain, Xtest, ytrain, ytest = train_test_split(X, Y, test_size=0.30, random_state=42)
print('X',X.shape)
print('Xtrain',Xtrain.shape)
print('Xtest',Xtest.shape)

X (1599, 11)
Xtrain (1119, 11)
Xtest (480, 11)


# Correlacion entre variables de entrada y Quality

In [43]:
Co = {}

for i in range (11):
    cols= Xtrain[:,i]
    corr= abs( np.corrcoef(cols,ytrain)[0,1] )
    Co[i]=corr
    print(corr)

print()
print("Correlaciones[]: ")
print(Co)
    


0.09920256591296486
0.3837827291004116
0.19697359834961337
0.006775665706756029
0.13116025872045628
0.05738283677446502
0.20158787409370493
0.18868024966005742
0.026849744732810796
0.24303697323454387
0.4920335659065441

Correlaciones[]: 
{0: 0.09920256591296486, 1: 0.3837827291004116, 2: 0.19697359834961337, 3: 0.006775665706756029, 4: 0.13116025872045628, 5: 0.05738283677446502, 6: 0.20158787409370493, 7: 0.18868024966005742, 8: 0.026849744732810796, 9: 0.24303697323454387, 10: 0.4920335659065441}


In [44]:
sort_co = sorted(Co.items(), key=lambda kv: kv[1]) # Orden Ascendete
sort_co =sort_co[::-1]                             # Orden Descendente 
sort_co

[(10, 0.4920335659065441),
 (1, 0.3837827291004116),
 (9, 0.24303697323454387),
 (6, 0.20158787409370493),
 (2, 0.19697359834961337),
 (7, 0.18868024966005742),
 (4, 0.13116025872045628),
 (0, 0.09920256591296486),
 (5, 0.05738283677446502),
 (8, 0.026849744732810796),
 (3, 0.006775665706756029)]

# Modelo MLP 

In [45]:
from sklearn.metrics import accuracy_score
MLP=[]
cont=1
for i in sort_co:
    MLP.append(i[0])
    model = MLPRegressor()
    model.fit(Xtrain[:,MLP],ytrain)
    ypred = model.predict(Xtest[:,MLP])
    
    
    print("MLP con: ", cont )
    print()
    print("Indices: ",MLP)
    print('Ytest',ytest[:10])
    print('Ypred',ypred[:10])
    print('MAE:',mean_absolute_error(ytest,ypred))
    print('MSE:',mean_squared_error(ytest,ypred))
    print('Y values:',min(Y),max(Y))
    print('Accuracy',accuracy_score(ytest,ypred.round()))
    print()
    cont+=1

    

MLP con:  1

Indices:  [10]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.19394568 5.1438882  5.34411809 5.29406062 5.69452039 5.09383073
 5.24400315 5.39417556 5.29406062 5.84469281]
MAE: 0.5744859174136769
MSE: 0.5510966987442306
Y values: 3.0 8.0
Accuracy 0.5354166666666667



C:\Users\oscar\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLP con:  2

Indices:  [10, 1]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.21244275 5.22871661 5.23485216 5.40376382 5.76201393 5.27551803
 5.27757573 5.16769903 5.4749948  5.74371992]
MAE: 0.5492190582051069
MSE: 0.47974526038438725
Y values: 3.0 8.0
Accuracy 0.5333333333333333



C:\Users\oscar\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLP con:  3

Indices:  [10, 1, 9]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.25322908 5.04620792 5.65596236 5.27152979 5.67606886 5.11802946
 5.22914927 5.0305549  5.55799    5.76064282]
MAE: 0.5322653059723891
MSE: 0.45059857977676265
Y values: 3.0 8.0
Accuracy 0.5291666666666667

MLP con:  4

Indices:  [10, 1, 9, 6]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.35494274 5.15438753 5.69658505 5.4192678  5.84714313 5.21853846
 5.34832148 5.11510591 5.66348975 5.90671   ]
MAE: 0.5352921947758523
MSE: 0.4491970248512282
Y values: 3.0 8.0
Accuracy 0.5625



C:\Users\oscar\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLP con:  5

Indices:  [10, 1, 9, 6, 2]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.30122249 5.09955787 5.76657709 5.34972364 5.7945353  5.19481026
 5.2970678  5.08423082 5.67887625 5.86570264]
MAE: 0.520558876695374
MSE: 0.4265590360066594
Y values: 3.0 8.0
Accuracy 0.5666666666666667



C:\Users\oscar\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLP con:  6

Indices:  [10, 1, 9, 6, 2, 7]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.29155887 5.05247963 5.69512067 5.3593804  5.78992777 5.16177712
 5.18395552 5.09698031 5.79298425 5.85545475]
MAE: 0.5139953090251725
MSE: 0.42037843890607907
Y values: 3.0 8.0
Accuracy 0.5541666666666667



C:\Users\oscar\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLP con:  7

Indices:  [10, 1, 9, 6, 2, 7, 4]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.16530153 4.96186515 5.69038819 5.21946516 5.75150922 5.05975379
 4.98748829 5.04416044 5.64815449 5.79841257]
MAE: 0.5088980777321438
MSE: 0.4210520510442031
Y values: 3.0 8.0
Accuracy 0.5604166666666667

MLP con:  8

Indices:  [10, 1, 9, 6, 2, 7, 4, 0]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.22998507 5.08628417 5.69709192 5.30613094 5.68486696 5.14047818
 5.07013543 4.89795699 5.67955356 5.62627961]
MAE: 0.5155742945886126
MSE: 0.43141464107348
Y values: 3.0 8.0
Accuracy 0.5625



C:\Users\oscar\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


MLP con:  9

Indices:  [10, 1, 9, 6, 2, 7, 4, 0, 5]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.09369743 4.94384957 5.4933093  5.37373882 5.70585237 4.91778941
 4.95912649 5.1131744  5.56868191 5.70900184]
MAE: 0.5127712194439666
MSE: 0.4389532010201626
Y values: 3.0 8.0
Accuracy 0.55625

MLP con:  10

Indices:  [10, 1, 9, 6, 2, 7, 4, 0, 5, 8]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.19140388 5.01264035 5.65516082 5.42326999 5.81386109 5.05219843
 4.94509901 5.18575667 5.51159057 5.8428205 ]
MAE: 0.5152403029128634
MSE: 0.4386589930350713
Y values: 3.0 8.0
Accuracy 0.5583333333333333

MLP con:  11

Indices:  [10, 1, 9, 6, 2, 7, 4, 0, 5, 8, 3]
Ytest [6. 5. 6. 5. 6. 5. 5. 5. 5. 6.]
Ypred [5.31936288 5.19628635 5.84357928 5.47034474 5.88364536 5.2121124
 5.0764184  5.18513737 5.80546865 5.85011724]
MAE: 0.5193716437344045
MSE: 0.4237373717118453
Y values: 3.0 8.0
Accuracy 0.5604166666666667



C:\Users\oscar\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


###  Indices:  [10, 1, 9, 6, 2]   (alcohol, volatile acidity, sulphates, total sulfur dioxide, citric acid) fueron los indicadores que aportaron mayor precisión con un: Accuracy 0.5666666666666667

# Modelo Decision Tree Regressor

In [49]:
DTR=[]
cont=1
for i in sort_co:
    DTR.append(i[0])
    model = DecisionTreeRegressor()
    model.fit(Xtrain[:,DTR],ytrain)
    ypred = model.predict(Xtest[:,DTR])
    print("DTR con: ", cont )
    print()
    print("Indices: ",DTR)
    print('MAE:',mean_absolute_error(ytest,ypred))
    print('MSE:',mean_squared_error(ytest,ypred))
    print('Y values:',min(Y),max(Y))
    print('Accuracy',accuracy_score(ytest,ypred.round()))
    print()
    cont+=1


DTR con:  1

Indices:  [10]
MAE: 0.6047472623180555
MSE: 0.5960486327303935
Y values: 3.0 8.0
Accuracy 0.5208333333333334

DTR con:  2

Indices:  [10, 1]
MAE: 0.5164136904761905
MSE: 0.6908126299130763
Y values: 3.0 8.0
Accuracy 0.56875

DTR con:  3

Indices:  [10, 1, 9]
MAE: 0.5526041666666667
MSE: 0.7514322916666667
Y values: 3.0 8.0
Accuracy 0.5416666666666666

DTR con:  4

Indices:  [10, 1, 9, 6]
MAE: 0.4979166666666667
MSE: 0.6770833333333334
Y values: 3.0 8.0
Accuracy 0.5854166666666667

DTR con:  5

Indices:  [10, 1, 9, 6, 2]
MAE: 0.5145833333333333
MSE: 0.7020833333333333
Y values: 3.0 8.0
Accuracy 0.5708333333333333

DTR con:  6

Indices:  [10, 1, 9, 6, 2, 7]
MAE: 0.5208333333333334
MSE: 0.7208333333333333
Y values: 3.0 8.0
Accuracy 0.56875

DTR con:  7

Indices:  [10, 1, 9, 6, 2, 7, 4]
MAE: 0.50625
MSE: 0.6979166666666666
Y values: 3.0 8.0
Accuracy 0.5854166666666667

DTR con:  8

Indices:  [10, 1, 9, 6, 2, 7, 4, 0]
MAE: 0.47708333333333336
MSE: 0.6395833333333333
Y values: 3

### Indices:  [10, 1, 9, 6, 2, 7, 4, 0, 5]  (alcohol, volatile acidity, sulphates, total sulfur dioxide, citric acid, density, chlorides, fixed acidity, free sulfur dioxide) fueron los indicadores con los que la precisión suele ser mas efectiva con un: Accuracy 0.6229166666666667